lib/connection.js

```diff
@@ -145,32 +151,56 @@ class Connection extends EventEmitter {
     }
 
     let message;
+    let isNewSession = true;
+    const handshakePayload = new Array(2);
+
+    if (login) {
+      if (!password) {
+        this.session = login;
+        handshakePayload[0] = 'session';
+        handshakePayload[1] = [this.session.id, this.session.receivedCount];
+        isNewSession = false;
+      } else {
+        handshakePayload[0] = 'login';
+        handshakePayload[1] = [login, password];
+      }
+    }
+
     if (version) {
-      message = login && password ?
-        this._createMessageWithArray(
-          'handshake', [name, version], 'login', [login, password]
-        ) :
-        this._createMessageWithArray(
-          'handshake', [name, version]
-        );
+      message = this._createMessageWithArray(
+        'handshake', [name, version], ...handshakePayload
+      );
     } else {
-      message = login && password ?
-        this._createMessage('handshake', name, 'login', [login, password]) :
-        this._createMessage('handshake', name);
+      message = this._createMessage('handshake', name, ...handshakePayload);
     }
 
     const messageId = message.handshake[0];
-    this._callbacks[messageId] = (error, sessionId) => {
-      if (login && password && !error) {
-        this.username = login;
-      }
-      this.sessionId = sessionId;
-      if (callback) {
-        callback(error, sessionId);
-      }
-    };
+    if (isNewSession) {
+      this._callbacks[messageId] = (error, sessionId) => {
+        if (!error) {
+          if (login && password) {
+            this.username = login;
```

Comment:
I think, it's better to duplicate it because it doesn't use much more memory but at the same time helps to avoid long access to it, like `connection.session.username`.